Lab1 - Generate Fibonacci Series

jimbosoft opened the pull request:
Fixes #630 

Review of colleague's PR #514 

#### Changes proposed in this PR:
Submit lab1 Fibonacci from the go course
## -

harirakr committed:
Add lab5 - stringer interface

Commit:
Implementation of Fibonacci Series for lab1

Modified version of the Excersise: Fibonacci closure from the Tour of Go
It already contains some review comments from a previous attempt

Commit:
Added handling of negative numbers

Commit:
Unit tests to verify all operations

Commit:
Added required README for documentation

willshen8 commented on 01_fib/jimbotech/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/.vscode/*
```
Only source files should be committed.

jimbosoft replied:
I usually check in the .gitignore and others also consider it normal practise as per https://stackoverflow.com/questions/5765645/should-you-commit-gitignore-into-the-git-repos, however in this context it might not add a lot, so I removed it.

willshen8 commented on 01_fib/jimbotech/README.md:
```diff
@@ -0,0 +1,15 @@
+
+
+
+> Written with [StackEdit](https://stackedit.io/).
```
Great to see you putting together a ReadME file, I don't think anyone else has done it. Perhaps remove the part that is auto-generated by the tools. e.g. 
`> Written with [StackEdit](https://stackedit.io/).`

jimbosoft replied:
StackEdit is an open source community and I feel I should acknowledge their efforts by crediting them in my work, so if it is ok, I think leaving in the credit is the right thing to do.

willshen8 replied:
Fair enough.

willshen8 commented on 01_fib/jimbotech/ex-fib_test.go:
```diff
@@ -0,0 +1,46 @@
+	"testing"
+)
+
+func TestFib(t *testing.T) {
```
It would be worthwhile to test the boundary cases such as `fib(0)` and `fib(1)` and even better larger fib number.

jimbosoft replied:
Done

willshen8 commented on 01_fib/jimbotech/ex-fibonacci.go:
```diff
@@ -0,0 +1,47 @@
+// in the fibonnacci sequence, starting at 1
+// and sets up the next which will be returned
+// in the subsequent call
+func fibonacci() func() int {
```
The tab takes up 8 character spaces here. I would think normally a tab/indentation would be 2 or 4 spaces.

jimbosoft replied:
I agree, I much prefer 4 characters, but after having spent a lot of time beating my editor into submission, I found "go fmt" simply puts the 8 space tab right back. I also checked other peoples submissions and they all use 8 space tabs. As much as I am agreeing with you, I have been defeated and will not be able to implement your suggestion.

willshen8 replied:
I think you are right, the editor shows up as normal indentation, and gofmt puts it back as 8.

willshen8 commented on 01_fib/jimbotech/ex-fibonacci.go:
```diff
@@ -0,0 +1,47 @@
+
+var out io.Writer = os.Stdout
+
+// fibonacci returns a function that returns
```
I can't comment on closure and it is definitely an interesting way of doing it. I personally feel if you have to write this much comment to explain what the function does, then it needs to be simplified. From a learning perspective, it might be easier to learn how it is done as using go structure and syntax.

jimbosoft replied:
"Closure" is a key concept in go as per the tour of go: https://tour.golang.org/moretypes/25 I added all the comments so in six months time I can still remember why I did it. Hope it is ok if I keep it just like this.

willshen8 commented on 01_fib/jimbotech/ex-fibonacci.go:
```diff
@@ -0,0 +1,47 @@
+	f := fibonacci()
+
+	if n < 0 {
+		for i := 0; i > n; i-- {
```
There is more idiomatic way of writing [for loops ](https://golang.org/doc/effective_go.html) and you have used already on line 14 of the test file.

jimbosoft replied:
In line 14 of my test I was "ranging" over a collection. This is not the case here. I am simply counting from 0 to n. I can't think of any other way to do it, unless you could show me how?

willshen8 reviewed (changes requested):
Apart from the comments below, I think the filename needs to be snakecase to align with convention:
https://stackoverflow.com/questions/35694137/file-name-convention-for-compound-words

willshen8 commented:
Also please open an 'issue' and put the issue number inside `Fixes #` of the PR description. The go-course require commit messages to be squashed, so you might have to do an interactive rebase.

willshen8 commented:
I moved the issue number up in the PR description (into the first line), I think maybe I wasn't clear. Hope you see it now.

nicko-lee committed:
Lab9 (#627)

* Implement lab9 requirements

Implement ability to convert between Go in memory objects to JSON
representation and vice versa.

Implement ability to accept user input from CLI and read data from
a file.

Load data from json text file and store in an instance of storer.

* Add tests to main_test.go

* Update README file to reflect changes in lab9

* Fix pesky scopelint error

* Fix pesky test coverage issue

Create new test with totally invalid json

* Fix final lint problem

Lint issue: should merge variable declaration with assignment on next line

* Add 2 json files to read into program

* Fix PR feedback

1) add defer f.close() to close file
2) use kingpin.ExistingFile no kingpin.String
3) tried using kingpin.fatalusage() to replace my printusage() func
   but struggled to hit it for test coverage
4) Remove ./ at start of paths
5) separate flag parsing and loading a file

camscale commented on 01_fib/jimbotech/ex_fibonacci.go:
```diff
@@ -0,0 +1,47 @@
+
+	if n < 0 {
+		for i := 0; i > n; i-- {
+			factor := int(math.Pow(-1, float64(i*-1)))
```
The computation of negative fibonacci numbers really belongs where the sequence is calculated. This function is responsible for printing `n` fib numbers. Instead you've split the computation over two separate functions for negative numbers. For instance, if you wanted to have another function return a slice of fib numbers instead of printing them, you would have to duplicate this negative number handling. But you've kind of backed yourself into a corner with your closure implementation of fib - it has no notion of which direction to go in so it cannot compute negative numbers.

Also the computation of -1 or 1 here is a bit obtuse. `i*-1` is more simply expressed as `-i` and using floating point calculations just to alternate between -1 and 1 seems unnecessary. If it were possible to write it as `-1^-i` in simple mathematical notation, then it may be worth doing. But with the casts and `math.Pow` instead of `^` or similar, it becomes that much harder to mentally parse and understand. An alternative is to use a variable `sign` initialised to `1`, and each time around the loop you just do `sign *= -1` - that way it alternates but is a much simpler expression.

jimbosoft replied:
Done ... however I feel hardly done by. I feel I am handling the negative numbers in the right place as the series is being generated there. So now I have separated the generation from the printing. Am I going to be criticised for inefficiency now? I was hoping for bonus points not bonus corrections. Secondly, as this is a exerciser I felt proud of my play with the math library and expected applause. Yes I did not like floating point either. When I re-wrote it I needed ABS, guess what, floating point again. So can I use any of the math library?

camscale commented on 01_fib/jimbotech/ex_fibonacci.go:
```diff
@@ -0,0 +1,47 @@
+			factor := int(math.Pow(-1, float64(i*-1)))
+			fmt.Fprintln(out, f()*factor)
+		}
+
```
nitpick: stray blank line. unnecessarily spaces out the code, which matters on real code that does not fit on one screen. blank lines to separate "paragraphs" of code are good, but do think about what your "code paragraphs" are.

jimbosoft replied:
Done

camscale commented on 01_fib/jimbotech/ex_fib_test.go:
```diff
@@ -0,0 +1,70 @@
+
+	for _, val := range fibResults {
+		if r := f(); r != val {
+			t.Errorf("Returned value %v does not match %v", r, val)
```
error messages in Go should start with lower case letters. This is because error messages are often wrapped as the propagate causing a SeQuEnCe of lower/upper case letters: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

jimbosoft replied:
Done

camscale commented on 01_fib/jimbotech/README.md:
```diff
@@ -0,0 +1,15 @@
+# Fibonacci Number Generator
+## Overview
+As part of the ANZ "go" course the first lab is to produce a piece of software that will produce the Fibonacci series for fib(7). It was also an exercise in the tour.golang.org.
```
nitpick: Markdown is an extension of plain old ascii documents, in that ascii markup is added so it can be prettier when rendered as markdown. But the files should still be readable as ascii documents. This means wrapping text around 80 chars (the general convention), as people use all sorts of different width terminals, and some that require side scrolling to read wider text. So please make markdown text readable as plain ascii as originally intended.

Plenty has been written about line length / text width - here's one example: https://baymard.com/blog/line-length-readability

jimbosoft replied:
Done

camscale reviewed (changes requested):
The file name `ex_fib_test.go` should be called `ex_fibonacci_test.go` - i.e. the same as the name as the file being tested but with `_test` appended before the extension. This is not something that is explicitly documented but is a convention I've never seen not followed.

My general approach to these reviews is to classify my comments into two types: Go-specific ones that need to be addressed to merge, and general software engineering practice ones which are advisory.

In this review, just the comments about error strings are Go-specific. So please at least fix those up (all instances - I did not leave a comment on every one, so look at all your error strings) before submitting for re-review. It is up to you as to whether you want to address the other comments, but it is good practice to at least respond to all comments with an acknowledgement or reason for ignoring.

Finally, a couple of comments on your commits and commit messages. Have a read of https://chris.beams.io/posts/git-commit/ (also linked from the slides) on what makes a good commit message. In particular, your first commit titled "INITIAL COMMIT" does not have a good title. It is shouty and non-descriptive. The initial split into five commits is fine and well done, but it does mean you need to take more action with your subsequent fixup commits - the should be squashed with the commits where the original code being fixed was written. We do not want to see fixup commits merged into master - the place to get your code and commits into order is the pull request. When merging it should look as though a series of commits was written "perfectly" the first time. If you started with one commit and intended to only keep one commit, you could do a "squash merge" when merging the PR to remove all evidence of the fix-ups. But if you want to retain your original split, you'll need to manually apply the fix-ups with `git rebase`. Remember, history is written by the victors. Be a victor!

Commit:
Add Lab 8 - Project Layout (#621)

Commit:
Fixed issues resulting from the last PR review

* Removed unwanted blank lines
* Separated the printing of the series from the generation
* Removed math function that use floating point

Commit:
Fixed test file as a result of the last PR review

* Renamed the file to go convention
* Made sure that error messages start with a lower case letter

Commit:
Fixed formatting as a result of the PR review

Commit:
Changed sign reversal

camscale reviewed (changes requested):
There is something very wrong with this PR. It has 29 files changed in it with code from other people's lab. I suggest you checkout the latest master and rebase on that, then force push. That *should* clean it up, but I am not sure how you got here so cannot offer precise guidance of how to get out.

I cannot re-review until this is fixed.

jimbosoft committed:
Merge branch 'master' into lab1

Commit:
Removed unwanted blank lines

camscale reviewed (approved):
LGTM.

BTW, if you wanted to incorporate negative fibonacci numbers into your closure way of generating numbers:
```
func fibonacci(direction int) func() int {
	if direction < 0 {
		direction = -1
	} else {
		direction = 1
	}
	secondLast, last := 0, 1

	return func() int {
		result := last
		secondLast, last = last, secondLast+(last*direction)
		return result
	}
}
```

Then you could just call `f := fibonacci(n)` and iterate over the result of `f()` which would remove the need for the intermediate slice. Or slightly more verbose but taking care of the iteration properly:
```
f := fibonacci(n)
limit := n
if n < 0 {
     limit = -n
}
for i := 0; i < limit; i++ {
    fmt.Fprintln(out, f())
}
```

I'm not requesting this change, just pointing out what I had in mind with my previous comments.

jimbosoft committed:
Merge branch 'master' into lab1
